2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+  -Module Az.KeyVault -Version 1.0.0
+
+# When using remote vaults, expectation is that secrets are
+# stored using their tooling, so this module is only for retrieving secrets
+# from remote vaults
+$azDevOpsToken = Get-Secret -Name AzDevOps
+Invoke-RestMethod https://dev.azure.com/… -Credential $azDevOpsToken -Authentication Basic
+```
+
+Registering and using local secret:
+
+```powershell
+# For local vault, we can register custom secrets
+# In this example, we store a PSCredential object
+Register-Secret -Secret $cred -Name MyCreds
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
+passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
+and the module version to `-ModuleVersion`.
+This can be done as part of importing the module or explicitly by the user.
+
+`Register-SecretsVaultExtension` has a `-Default` switch to register the
+default local vault (if applicable, see above) if it has been unregistered.
+
+`Register-SecretsVaultExtension` takes a `-AccessToken` or `-Credential` parameter
+for authenticating with the remote vault.
+If neither are supplied, then the current user context is used and relies on
+the extension cmdlet to handle authentication.
+
+`Unregister-SecretsVaultExtension` should be used to remove any existing
+registration or when updating a registration to a new version.
+
+`Get-SecretsVaultExtension` will enumerate registered extensions returning
+the module name, module version, and cmdlet used.
+
+### Storing Secrets
+
+The `Register-Secret` cmdlet is used to store a secret.
+The `-Name` must be unique across all secrets regardless if local or remote.
+A `-Secret` parameter accepts an object that can be either a PSCredential
+or SecureString.
+
+### Retrieving Secrets
+
+The `Get-Secret` cmdlet is used to retrieve secrets.
```

Comment:
I recommend using Get-SecretValue rather than Get-Secret. This will allow you to support metadata operations with Get-Secret (e.g. list all secrets, retrieve secret metadata etc).

For local secrets this isn't particularly important, but for remote stores it's critical, since that's how you'll view/manage permissions, actions, rotation settings etc.

Reply:
I think its useful to browse both local and remote stores. When I change my password, I have a script that gets the list of Windows resources (mapped drives, servers) stored in Credential Manager so that I can update the passwords.

Reply:
@thedevopsmachine I don't think the purpose of this is be a full blown cross-service secrets management solution and I don't think metadata should be part of it. For that, users and developers should fall back to native implementations. I think the purpose for this is more of cross-platform simple secret retrieval so that secrets do not have to be stored in plain text and so that developers o not have to reinvent multiple password retrieval solutions per OS/environment/company

Reply:
I agree, and I don't think things like permissions management should be in scope. However, being able to list what secrets you have in a vault seems like a pretty fundamental capability. Since listing available secrets and retrieving secret values often require different permissions, it doesn't make sense to couple them into a single "Get-Secret" cmdlet. Ideally there would be a "List" verb but there isn't, so we have to do the separation in the noun.

If you want to use Get-Secret for the value and Get-SecretList for enumeration then that works too. I just think it's important to decouple listing and retrieving.

Reply:
The `Get` verb is typical used with a single noun for a list and a single item and it's generaly not good practice to separate them. 

But again, I'm not sure getting a list of secrets makes sense. The motivation of this feature is not managing secrets (I think the RFC title is misleading). It's not proposing a solution to manage secrets in the sense that ops and infra people would be familiar with. This is more about simple store and retrieve operations for secrets used in PowerShell. It's intent is not, for example, to maintain the secrets in your Azure Key Vault for 3rd party APIs your C# Azure Functions access.

Reply:
"Generally not good practice to separate them" - is it though? Then why are Get-Item and Get-ChildItem different Cmdlets? There's also Get-ItemProperty and Get-ItemPropertyValue. Both are examples of collections vs single items, and both are in the base Cmdlets for PowerShell.

I'm not suggesting it should manage secrets in detail. I'm arguing two things:

- Listing of secrets should be included in the definition of "store and retrieve operations". If it *must* be in the same Cmdlet as retrieving the secret, then the two operations should be isolated by parameter sets. 
- Metadata should be included in the Get-Secret response (i.e. not just a SecureString, PSCredential, or plaintext string). All the secret management solutions I've worked with will return metadata alongside the secret. That information can be useful. Perhaps it's the local admin password of the machine, and I want to know the last time it was updated, so that if it's past a certain amount of time, I can rotate it. Or maybe I want to Write-Warning that it's about to expire. Since the remote provider *has* that information in the response, it seems unnecessarily limiting to strip it out.

Reply:
RE: `Get-Item` vs `Get-ChildItem` -- if we only had `Get-Item`, you _have_ to use `Get-Item C:\*` to list child items (yay asterisks, how fun doing that literally every time). Or you have to introduce a "somewhat Recurse" parameter, or use `-Recurse -Depth 1` -- all of which are incredibly unsightly for what is basically a default operation in a lot of OSes (e.g., `ls` and `dir` _both_ default to listing items in a given directory). They exist because there is a distinction between "i want to get this specific container item" and "i want to get everything _in_ this specific container item".

Unless we're actually looking at having subfolders upon subfolders of secrets, a single cmdlet where the default is to list makes the most sense. If you want a single secret, you'd _target_ that secret specifically. If you're not targeting, listing makes sense. See also: most other `Get-*` cmdlets, e.g., `Get-Command`, `Get-Process`, `Get-Service`, etc., etc.

Reply:
"Unless we're actually looking at having subfolders upon subfolders of secrets" - yup that's exactly what this needs to support if you want to support Remote Providers. A Remote Provider is a service dedicated to secret management, so it's naturally going to contain lots of secrets. When you have a lot of secrets, to make them manageable you will use either folders or prefixes (depending on the provider).

I'm fine with Get-Secret being able to do both list and single value, but the concept of folders or prefixes should be supported. For example:

### Retrieving a secret
```
Get-Secret -Id 'production.webservers.iisuser'
or
Get-Secret -Path 'production.webservers.iisuser'
```

### Listing secrets
```
Get-Secret -Prefix 'production.webservers' -MaxItems 5
```

Reply:
If this were going to be a provider-based model we'd just be using `Get-ChildItem Secret:` 

@SteveL-MSFT is this going to be a PSProvider or a separate implementation?

Reply:
FWIW, I don't believe that calling it Get-Secret stops it from having metadata operations.  There are cmdlets that act differently depending on the parameters use (think Get-EventLog/ Get-EventLog -ListLog , Get-Counter / Get-Counter -ListSet)

Reply:
@vexx32 PSProvider is out of scope, but could be added in the future.  As @markekraus noted, this feature (at least in this iteration) is not intended to be a complete abstraction for management of secrets, but simplify and standardize access to secrets.  I do think it would be common for users to want to list secrets as they are doing authoring of their scripts.  I was thinking of following the `Get-Module` model where `Get-` without a name does enumerate the secrets (just the name and vault, not the secret itself).  As for folders, users can namespace it themselves with the name and do post filtering using normal PowerShell pipeline filtering.  If it is common that enumerating and retrieving secrets requires different permissions, we'll have to figure out a way to accommodate that.  For now, I'm leaving it as a problem to the extension to solve.